openhtf/core/phase_executor.py

```diff
@@ -251,6 +261,13 @@ def stop(self):
     It will raise a ThreadTerminationError, which will cause the test to stop
     executing and terminate with an ERROR state.
     """
+    self.test_state.logger.debug('Stopping PhaseExecutor')
     current_phase_thread = self._current_phase_thread
     if current_phase_thread:
-      current_phase_thread.kill()
+      kill_start = time.time()
+      current_phase_thread.stop()
+      while current_phase_thread.is_alive():
+        time.sleep(.1)
+      self.test_state.logger.debug('Took %.3f seconds to kill %s',
+          time.time() - kill_start,
+          current_phase_thread.name)
```

Comment:
I definitely agree a singular place is better than the 3 places.  :)

Short term:  We should fix this race condition.   We are relying upon the current_phase_thread to actually be killed in a timely fashion and return the PhaseOutcome(ThreadTerminationError) inside of phase_executor.execute_phases for it to correctly unwind things.  Otherwise things fall apart in unpredictable fashion.

Longer term: moving it to test_executor would probably get my vote.  There is some logic (that I haven't quite groked) around the finalize that we'd have to be careful about.